### PR TITLE
Support multiple DNS queries

### PR DESCRIPTION
**Pull request by davydog187**
This PR extends the API to allow for a list of DNS queries. This is useful if you have multiple services that you want to cluster together

### CODE REVIEW

File: lib/dns_cluster.ex
Comment (davydog187) [inline lib/dns_cluster.ex]: note that we don't validate that the list contains only strings. Happy to add that if we think its necessary (seems like its not to me)

**Comment by davydog187** [comment]
This PR needs a little more thought due to the basename restriction on connected nodes. I'll try to give this more time tomorrow

**Review by josevalim — APPROVED** [review]

**Comment by joshk** [comment]
This is really awesome! I was going to create a similar PR but was happy to see it was already done!

Would it be possible to accept a list of domains separated by commas? 
eg. `"myapp.internal"` or `["foo.internal", "bar.internal"]` or `"foo.internal,bar.internal"`

This would make it a lot easier to configure via environment variables. 
eg.`DNS_CLUSTER_QUERY="foo.internal,bar.internal"`

**Comment by davydog187** [comment]
hello @joshk! I still need to make some tweaks to this PR before it can be merged, unfortunately.

> Would it be possible to accept a list of domains separated by commas?
eg. "myapp.internal" or ["foo.internal", "bar.internal"] or "foo.internal,bar.internal"

It is certainly possible, although I don't recommend introducing this to the API of this library. While not explicitly mentioned in the [anti-patterns](https://hexdocs.pm/elixir/1.16.0-rc.0/design-anti-patterns.html#using-application-configuration-for-libraries) guides, I believe it is trivial to do the string splitting in your application code. e.g.

```elixir
# config/runtime.exs
config :my_app, dns_cluster_query: String.split(System.get_env("DNS_CLUSTER_QUERY"), ",")

# lib/application.ex
{DNSCluster, query: Application.get_env(:my_app, dns_cluster_query)}
```

I think this keeps the library API simple while enabling maximum flexibility for application developers using this library

**Comment by davydog187** [comment]
Also, I suppose that it is valid to have a `,` in your DNS query, so that's another reason not to support it

**Comment by davydog187** [comment]
@josevalim @chrismccord I have updated this PR to add support for connecting to nodes with a basename different than the host. To support this, I've extended the API to allow the user to specify a query, a `{basename, query}` or a list of either.

I've updated the documentation to reflect this change, please let me know if you want to go a different direction. I'm not an expert in erlang distribution, but I'm not aware of another parameter that would warrant a more complicated datastructure. If that is the case, we should probably use a map or struct instead.

**Comment by davydog187** [comment]
I'm running this branch in production and it works great!

Screenshot 2024-01-13 at 9 32 34 AM

**Comment by davydog187** [comment]
I'm running this branch in production and it works great! (Redacted IPs)

```elixir
iex> Node.list()
[
    :"sauron-prod@fdaa:1:aaaaaaaaaaaaaaaaaaaaa:2",
    :"leaderboard-prod@fdaa:1:aaaaaaaaaaaaaaaaaaaaaa:2",
    :"leaderboard-prod@fdaa:1:aaaaaaaaaaaaaaaaaaaaaa:2"
]
```

File: lib/dns_cluster.ex
Comment (davydog187) [inline lib/dns_cluster.ex]: This allows nested lists which is not intended

**Comment by davydog187** [comment]
For anyone interested in this feature, I encourage you to test it out! I've been running it in production for two weeks now without any issue

**Comment by schrockwell** [comment]
@davydog187 Thanks, I'm trying this on Fly right now. To set it up, I had to do a couple things:

1. In `rel/env.sh.eex`, set `export RELEASE_NODE="${FLY_APP_NAME}@${FLY_PRIVATE_IP}"` (note that `$FLY_IMAGE_REF` has been removed from the node basename)
2. Parse the `$DNS_CLUSTER_QUERIES` which is in the format `"app1@app1.internal app2@app2.internal"`:
```elixir
dns_cluster_queries =
  System.get_env("DNS_CLUSTER_QUERIES", "")
  |> String.split(" ", trim: true)
  |> Enum.map(fn query ->
    case String.split(query, "@") do
      [basename, domain] -> {basename, domain}
      [domain] -> domain
    end
  end)
```
3. Set `DNS_CLUSTER_QUERIES` in both apps' TOML configs and deploy.

**Comment by davydog187** [comment]
Hey @schrockwell 👋🏼 

That all sounds well and good, however, I would personally just hardcode the queries in your TOML file, you shouldn't need to parse them given that they should be static per environment

**Comment by davydog187** [comment]
Sorry to ping you @chrismccord but is there anything holding up this PR besides being a low priority?

**Comment by yordis** [comment]
Hey folks, any opportunity to get this one in? 🙏🏻

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**
